src/main/python/ayab/utils.py

```diff
@@ -19,43 +19,50 @@
 #    https://github.com/AllYarnsAreBeautiful/ayab-desktop
 """Functions that do not have a parent class."""
 
+from __future__ import annotations
 import logging
+from typing import Any, Callable, Literal, Optional, TypeAlias, cast, TYPE_CHECKING
 import numpy as np
+import numpy.typing as npt
 import serial.tools.list_ports
 import requests
 
-from PySide6.QtWidgets import QMessageBox
+from PySide6.QtWidgets import QMessageBox,QWidget,QComboBox
 
 
-def even(x):
+def even(x:int)->bool:
     return x % 2 == 0
 
 
-def odd(x):
+def odd(x:int)->bool:
     return x % 2 == 1
 
 
-def display_blocking_popup(message="", message_type="info"):
+MessageTypes: TypeAlias = Literal["error","info","question","warning"]
+QMessageBoxFunc: TypeAlias = Callable[[Optional[QWidget],str,str,QMessageBox.StandardButton,QMessageBox.StandardButton],QMessageBox.StandardButton] 
```

Comment:
It irks me these aren't using inheritance or anything so I had to make this big typealias

Reply:
Might be worth fixing that.

Reply:
It's coming from Qt so not much we can do there. Not a real issue, just a minor inconvenience to typing.

Reply:
I made it less nonsense